components/TextFields/src/MDCTextInputController.m

```diff
@@ -0,0 +1,1201 @@
+      default:
+        underlineColor = normalColor;
+        underlineHeight = MDCTextInputUnderlineNormalHeight;
+        break;
+    }
+
+    self.textInput.underline.color =
+        (self.isDisplayingCharacterCountError || self.isDisplayingErrorText) ? self.errorColor
+                                                                             : underlineColor;
+    self.textInput.underline.lineHeight = underlineHeight;
+  }
+}
+
+#pragma mark - Underline Labels Fonts
+
++ (UIFont *)placeholderFont {
+  return [UIFont mdc_preferredFontForMaterialTextStyle:MDCFontTextStyleBody1];
+}
+
++ (UIFont *)underlineLabelsFont {
+  return [UIFont mdc_preferredFontForMaterialTextStyle:MDCFontTextStyleCaption];
+}
+
+#pragma mark - Properties Implementation
+
+- (void)setCharacterCountViewMode:(UITextFieldViewMode)characterCountViewMode {
+  if (_characterCountViewMode != characterCountViewMode) {
+    _characterCountViewMode = characterCountViewMode;
+
+    [self updateLayout];
+  }
+}
+
+- (void)setErrorAccessibilityValue:(NSString *)errorAccessibilityValue {
+  _errorAccessibilityValue = [errorAccessibilityValue copy];
+}
+
+- (void)setErrorColor:(UIColor *)errorColor {
+  if (![_errorColor isEqual:errorColor]) {
+    _errorColor = errorColor;
+    if (self.isDisplayingErrorText) {
+      [self updateLeadingUnderlineLabel];
+      [self updateTrailingUnderlineLabel];
+      [self updateUnderline];
+    }
+  }
+}
+
+- (void)setErrorText:(NSString *)errorText {
+  _errorText = [errorText copy];
+}
+
+- (void)setFloatingPlaceholderColor:(UIColor *)floatingPlaceholderColor {
+  if (![_floatingPlaceholderColor isEqual:floatingPlaceholderColor]) {
+    _floatingPlaceholderColor = floatingPlaceholderColor;
+    [self updatePlaceholder];
+  }
+}
+
+- (void)setFloatingPlaceholderScale:(NSNumber *)floatingPlaceholderScale {
+  if (![_floatingPlaceholderScale isEqualToNumber:floatingPlaceholderScale]) {
+    _floatingPlaceholderScale = floatingPlaceholderScale;
+    [self updatePlaceholder];
+  }
+}
+
+- (void)setHelperText:(NSString *)helperText {
+  if (self.isDisplayingErrorText) {
+    self.previousLeadingText = [helperText copy];
+  } else {
+    if (![self.textInput.leadingUnderlineLabel.text isEqualToString:helperText]) {
+      self.textInput.leadingUnderlineLabel.text = helperText;
+      [self updateLayout];
+    }
+  }
+}
+
+- (NSString *)helperText {
+  if (self.isDisplayingErrorText) {
+    return self.previousLeadingText;
+  } else {
+    return self.textInput.leadingUnderlineLabel.text;
+  }
+}
+
+- (void)setInlinePlaceholderColor:(UIColor *)inlinePlaceholderColor {
+  if (![_inlinePlaceholderColor isEqual:inlinePlaceholderColor]) {
+    _inlinePlaceholderColor = inlinePlaceholderColor;
+    [self updatePlaceholder];
+  }
+}
+
+- (BOOL)isDisplayingCharacterCountError {
+  return self.characterCountMax && [self characterCount] > self.characterCountMax;
+}
+
+- (BOOL)isDisplayingErrorText {
+  return self.errorText != nil;
```

Comment:
Nil. We support passing an empty string. That would set it to error state with no error text. For instance, when the character count is exceeded: The underline goes red, the character count goes red, but you don't necessarily have an error text.